darwin/pytorch/utils.py

```diff
@@ -0,0 +1,288 @@
+        return path
+    elif isinstance(polygon[0], dict):
+        for e in polygon:
+            path.append(e['x'])
+            path.append(e['y'])
+        return path
+    else:
+        return polygon
+
+
+def polygon_area(x, y):
+    '''
+    Returns the area of the input polygon, represented with two numpy arrays
+    for x and y coordinates.
+    '''
+    return 0.5 * np.abs(np.dot(x, np.roll(y, 1)) - np.dot(y, np.roll(x, 1)))
+
+
+def collate_fn(batch):
+    return tuple(zip(*batch))
+
+
+def extract_classes(files: List):
+    '''
+    Given a list of GT json files, extracts all classes and an mapping image index to classes
+
+    Input:
+        files: list of json files with the GT information of each image
+
+    Output:
+        classes: list of classes in the GT
+        idx_to_classes: mapping image index to classes that can be used to know all the classes
+                        in a given image index
+    '''
+    classes = {}
+    idx_to_classes = {}
+    for i, fname in enumerate(files):
+        with open(fname) as f:
+            d = json.load(f)
+            idx_to_classes[i] = []
+            for a in d["annotations"]:
+                cls_name = a["name"]
+                try:
+                    classes[cls_name].add(i)
+                except KeyError:
+                    classes[cls_name] = set([i])
+                if cls_name not in idx_to_classes[i]:
+                    idx_to_classes[i].append(cls_name)
+    return classes, idx_to_classes
+
+
+def fetch_darwin_dataset(
+    dataset_name: str,
+    client: Optional[Client] = None,
+    val_percentage: Optional[float] = 0.1,
+    test_percentage: Optional[float] = 0,
+    image_status: Optional[str] = "done",
+    force_fetching: Optional[bool] = False,
+    force_resplit: Optional[bool] = False,
+    split_seed: Optional[int] = None
+):
+    '''
+    Pull locally a dataset from Darwin (if needed) and create splits for
+    train, validation, and test.
+
+    Input:
+        dataset_name: name of the dataset in Darwin
+        client: Darwin client
+        val_percentage: percentage of images used in the validation set
+        test_percentage: percentage of images used in the test set
+        image_status: only pull images with under this status
+        force_fetching: discard local dataset and pull again from Darwin
+        force_resplit: discard previous split and create a new one
+        split_seed: fix seed for random split creation
+
+    Output:
+        root: local path to the dataset
+        split_path: relative path to the selected train/val/test split
+    '''
+    if client is None:
+        client = Client.default()
+
+    # Get data
+    dataset = None
+    local_datasets = {dataset.slug: dataset for dataset in client.list_local_datasets()}
+    if dataset_name in local_datasets:
+        if force_fetching:
+            # Remove the local copy of the dataset
+            dbpath = os.path.join(client.project_dir,  dataset_name)
```

Comment:
Minor comment: we tend to prefer `pathlib.Path` rather than `os.path` overall.

Reply:
Fixed :)